distributed/scheduler.py

```diff
@@ -282,78 +283,188 @@ class WorkerState:
 
     # XXX need a state field to signal active/removed?
 
+    _actors: set
+    _address: object
+    _bandwidth: int
```

Comment:
[From]( https://github.com/dask/distributed/pull/4302#discussion_r535325646 ) @mrocklin:

> Sometimes we use int and sometimes Py_ssize_t, can you help explain why we choose one over the other? From a readability perspective I would prefer int everywhere. I'm guessing that this varies a bit on 32bit systems?

Reply:
Oops this should be `Py_ssize_t`. Thanks for catching that.

<hr>

Unfortunately Cython treats [`int` as a Python object]( https://cython.readthedocs.io/en/latest/src/tutorial/pure.html#pep-484-type-annotations ):

> Note the use of cython.int rather than int - Cython does not translate an int annotation to a C integer by default since the behaviour can be quite different with respect to overflow and division.

For now I've just used `Py_ssize_t` as it is the largest signed integer available. So we are unlikely to run into issues with it. Though yeah this may vary between 32-bit and 64-bit systems. Figure if it is important to restrict this to a particular sized integer (like `int16`) we can do that in follow up work.

In general haven't been too focused on typing everything perfectly, but have instead tried to provide some type for everything and try to avoid `object` when possible. The goal being to generally get types across as much of the code as possible and tune afterwards.

Reply:
Fixed in latest comment

Reply:
Reverted and marked unresolved. Now am remembering why I used `int`. It seems like this should be an `int`, but sometimes it is a `float`.

For example, [the `Scheduler.bandwidth` is an `int` to start]( https://github.com/dask/distributed/blob/dcdb4651eb0331bb99caf3362c84d4878d94ba5b/distributed/scheduler.py#L1204 ). Though [`Scheduler.bandwidth_workers` contains `float`s]( https://github.com/dask/distributed/blob/dcdb4651eb0331bb99caf3362c84d4878d94ba5b/distributed/scheduler.py#L1205 ). Later on [the `Scheduler.bandwidth` becomes a `float`]( https://github.com/dask/distributed/blob/dcdb4651eb0331bb99caf3362c84d4878d94ba5b/distributed/scheduler.py#L1712-L1714 ).

Though the `float` value of [`Scheduler.bandwidth` infects values like `TaskState.nbytes`]( https://github.com/dask/distributed/blob/b7a2274f7ee8ac6a370604fbdd538c7a56d447af/distributed/tests/test_steal.py#L496 ), which isn't good.

Could use some guidance here. Do we expect `bandwidth` to be an `int`? If so, we should clamp down on some of these cases where it becomes a `float` by rounding it somehow and coercing it back to an `int`. If not, we should be more careful about how we mix `bandwidth` with other variables to ensure we get an `int` where we expect one.

Reply:
Submitted PR ( https://github.com/dask/distributed/pull/4311 ) for that specific issue.